pkg/cli/image/strategy/simplelookup.go

```diff
@@ -0,0 +1,166 @@
+package strategy
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+
+	operatorv1alpha1 "github.com/openshift/api/operator/v1alpha1"
+	operatorv1alpha1scheme "github.com/openshift/client-go/operator/clientset/versioned/scheme"
+	operatorv1alpha1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1alpha1"
+	reference "github.com/openshift/library-go/pkg/image/reference"
+	"github.com/openshift/library-go/pkg/image/registryclient"
+)
+
+// simpleLookupICSP holds ImageContentSourcePolicy variables to look up image sources
+// satisfies *Context AlternativeBlobSourceStrategy interface
+type simpleLookupICSP struct {
+	lock sync.Mutex
+
+	alternates []reference.DockerImageReference
+	icspFile   string
+	icspClient operatorv1alpha1client.ImageContentSourcePolicyInterface
+}
+
+func NewSimpleLookupICSPStrategy(file string, client operatorv1alpha1client.ImageContentSourcePolicyInterface) registryclient.AlternateBlobSourceStrategy {
+	return &simpleLookupICSP{
+		icspFile:   file,
+		icspClient: client,
+		alternates: nil,
+	}
+}
+
+func (s *simpleLookupICSP) FirstRequest(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	return nil, nil
+}
+
+// OnFailure returns a list of possible image references for the locator image reference, gathered from image content source policies, or an error
+func (s *simpleLookupICSP) OnFailure(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if err := s.resolve(ctx, locator); err != nil {
+		return nil, err
+	}
+	if len(s.alternates) == 0 {
```

Comment:
Having the resolve method set a global is weird (just return the list).  Also, why is this a list?  If two calls are made to the same strategy with different locators, then the strategy has to get a new alternates list.  So if you want to cache, you have to cache by locator, and you probably need to maintain the size of the cache (at least bound it to something high like 1024 entries).

In general, for this strategy you want to add the locator as the last option, not the first (also if it's not duplicated) when you're handling the "on error" strategy.

Reply:
Fixed. 

> In general, for this strategy you want to add the locator as the last option, not the first (also if it's not duplicated) when you're handling the "on error" strategy.

I'm not quite sure I understand what you mean by that?